src/shared/directives/conditional-show/services/condition-parser.service.ts

```diff
@@ -0,0 +1,204 @@
+import { _ as _score } from 'underscore';
+import peg from './condition.peg';
+import { FieldsUtils } from '../../../services/fields/fields.utils';
+import { ShowCondition } from '../../conditional-show/domain/conditional-show.model';
+
+export class ConditionParser {
+
+  /**
+   * Parse the raw formula and output structured condition data
+   * that can be used in evaluating show/hide logic
+   * @param condition raw formula e.g. TextField = "Hello"
+   */
+  public static parse(condition: string): any {
+    if (!condition) { return null };
+    condition = condition.replace(/CONTAINS/g, ' CONTAINS');
+    return peg.parse(condition.trim(), {});
+  }
+
+  /**
+   * Evaluate the current fields against the conditions
+   * @param fields the current page fields and their value
+   * @param conditions The PegJS formula output
+   */
+  public static evaluate(fields: any, conditions: any[], path?: string): boolean {
+    if (!conditions) { return true };
+    const validJoinComparators = ['AND', 'OR'];
+
+    const result: boolean = conditions.reduce((accumulator: boolean, condition, index: number) => {
+      const isJoinComparator = (comparator: string): boolean =>
+        (typeof comparator === 'string' && validJoinComparators.indexOf(comparator) !== -1);
+
+      if (isJoinComparator(condition)) { return accumulator };
+
+      let currentConditionResult = true;
+
+      if (Array.isArray(condition)) {
+        currentConditionResult = this.evaluate(fields, condition);
+
+        if (isJoinComparator(conditions[index - 1])) {
+          return this.evaluateJoin(accumulator, conditions[index - 1], currentConditionResult);
+        }
+      }
+
+      if (condition.comparator) {
+        const formula = condition.fieldReference + condition.comparator + condition.value;
+        const [field, conditionSeparator] = this.getField(formula);
+        const [head, ...tail] = field.split('.');
+        const currentValue = this.findValueForComplexCondition(fields, head, tail, path);
+        const expectedValue = this.unquoted(formula.split(conditionSeparator)[1]);
+        if (conditionSeparator === ShowCondition.CONTAINS) {
+          currentConditionResult = this.checkValueContains(expectedValue, currentValue);
+        } else {
+          currentConditionResult = this.checkValueEquals(expectedValue, currentValue, conditionSeparator);
+        }
+      }
+
+      if (isJoinComparator(conditions[index - 1])) {
```

Comment:
can this 4 ifs change to switch/case statements and breakdown to 4 separate methods

Reply:
it's bit complicated; but what I did is moved some code to new function.